torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,334 @@
+        """
+        hidden_states = []
+
+        # parse inputs
+        if images.ndim == 4:
+            n_tiles = 1
+            bsz, nch, w, h = images.shape
+            assert (
+                aspect_ratio is None or aspect_ratio.max() == 1
+            ), f"since images.ndim == 4, aspect ratio must be none or 1x1, got {aspect_ratio}"
+
+        elif images.ndim == 5:
+            bsz, n_tiles, nch, w, h = images.shape
+
+        else:
+            raise ValueError(
+                f"Unsupported number of dimensions: {images.ndim}. Expected 4 or 5."
+            )
+
+        # if aspect_ratio is not provided, it defaults to one tile [1,1]
+        if aspect_ratio is None:
+            aspect_ratio = torch.tensor([[1, 1]] * bsz, device=images.device)
+
+        images = images.reshape(bsz * n_tiles, nch, w, h)
+        aspect_ratio = aspect_ratio.reshape(bsz, 2)
+
+        # patch embeddings (tokens)
+        # out: (bsz, embed_dim, patch_grid_size, patch_grid_size)
+        x = self.conv(images)
+
+        # out: (bsz, patch_grid_size**2, embed_dim)
+        x = x.flatten(start_dim=2).permute(0, 2, 1)
+
+        _, n_tokens, embed_dim = x.shape
+
+        # pre_tile_pos_embed
+        if self.pre_tile_pos_embed:
+            x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+            x = self.pre_tile_pos_embed(x, aspect_ratio)
+
+        # apply cls token
+        x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
+        x = self.cls_token_embedding(x)
+        n_tokens += 1
+
+        # token_pos_embedding
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+        x = self.token_pos_embedding(x, aspect_ratio)
+
+        # norm
+        x = self.ln_pre(x)
+
+        # transformer with optional hidden layer outputs
+        x = x.view(bsz, n_tiles * n_tokens, embed_dim)
+        for layer_idx, transformer_layer in enumerate(self.transformer_layers):
+            if self.indices_return_hidden and layer_idx in self.indices_return_hidden:
+                hidden_states.append(x)
+            x = transformer_layer(x)
+
+        # norm
+        x = self.ln_post(x)
+
+        # reshape outputs
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+
+        if self.indices_return_hidden:
+            hidden_states = torch.stack(hidden_states, dim=-1)
+            hidden_states = hidden_states.reshape(
+                bsz, n_tiles, n_tokens, embed_dim, len(self.indices_return_hidden)
+            )
+        else:
+            hidden_states = torch.empty(0)  # dummy tensor
+
+        # post_tile_pos_embed
+        if self.post_tile_pos_embed:
+            x = self.post_tile_pos_embed(x, aspect_ratio)
+
+        # cls token projection
+        if self.cls_projection:
+            x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
+            x = self.cls_projection(x)
+            x = x.reshape(
+                bsz, n_tiles, 1, -1
+            )  # (bsz, n_tiles, num_tokens, cls_output_dim)
```

Comment:
Isn't it actually this?
```suggestion
            )  # (bsz, n_tiles, 1, cls_output_dim)
```

Also this is one reason why I don't like the optional CLS projection. It kinda makes the contract unclear -- for some versions of VisionTransformer I will get back all tokens, but for others I will only get back the CLS representation

Reply:
Is there a scenario where we don't add the CLS token and projection? Like it's ok to assume we add this token and then make it so that the user slices to remove the CLS token if they dont need it. We just need to fix the functionality - ok to not stuff a lot of generality into code. Of course I"m not an expert so disregard if this doesnt make sense

Reply:
In the implementations I have seen, we always add the CLS token. However, Flamingo has its own type of projection, and the CLS projection is not used, so we had to make it optional in the ViT.

We could think of both flamingo_projection and CLS_projection as adapters.

I chatted with Philip about it.
option 1: Remove CLS projection from ViT and make it an adapter. 
Cons: since CLS projection is part of the VIT, this would be confusing for users looking for ViT.

Option 2: Make the ViT accept any type of adapter, including CLS projection. If we pass the flamingo adapter, this becomes flamingo.
Cons: Not sure. Maybe the ViT + adapter distinction is not super clear;

Option 3: Leave CLS projection there, and add flamingo adapter logic in a different module
```
class FlamingoVisionEncoder():
     def __init__(vit, flamingo_adapter):
             ...
      def forward(images, aspect_ratio):
             x, hidden_states = vit(images, aspect_ratio)
             x = flamingo_adapter(x)
             return x
```

We picked option 3